File: android/app/src/main/java/com/myvu/client/ai/AiHttpClient.java
package com.myvu.client.ai;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * Shared plumbing for the provider clients: one JSON POST, one JSON answer.
 * Subclasses supply only what actually differs -- endpoint, auth headers,
 * request body and where the answer sits in the response.
 *
 * Non-streaming on purpose: the glasses' flow needs the whole answer up front,
 * because playback is gated between code:6 playState 1 and 2 -- there is nothing
 * useful to do with partial tokens.
 *
 * Blocking; must be called off the connection thread.
 */
public abstract class AiHttpClient implements AiClient {

    private static final int TIMEOUT_MS = 30000;

    protected final AiProvider provider;
    protected final String apiKey;
    /** Never blank: a blank Settings override falls back to the provider default. */
    protected final String model;
    protected final String systemPrompt;
    protected final boolean ignoreSsl;

    protected AiHttpClient(AiProvider provider, String apiKey, String model,
                           String systemPrompt) {
        this(provider, apiKey, model, systemPrompt, false);
    }

    protected AiHttpClient(AiProvider provider, String apiKey, String model,
                           String systemPrompt, boolean ignoreSsl) {
        this.provider = provider;
        this.apiKey = apiKey;
        this.model = (model == null || model.trim().isEmpty())
                ? provider.defaultModel : model.trim();
        this.systemPrompt = (systemPrompt == null || systemPrompt.trim().isEmpty())
                ? DEFAULT_SYSTEM_PROMPT : systemPrompt.trim();
        this.ignoreSsl = ignoreSsl;
    }

    @Override
    public boolean isConfigured() {
        return apiKey != null && !apiKey.trim().isEmpty();
    }

    /** Full URL to POST to. {@link #model} is set, for APIs that put it in the path. */
    protected abstract String endpoint();

    /** Sets the provider's auth header(s) on the request. */
    protected abstract void authorize(HttpURLConnection conn);

    /** The request JSON: model, system prompt and the one user question. */
    protected abstract String buildBody(String question) throws JSONException;

    /** Digs the answer text out of a 2xx response body. */
    protected abstract String extractText(String response) throws JSONException;

    @Override
    public String ask(String question) throws IOException {
        if (!isConfigured()) {
            throw new IOException(provider.label + " is not fully configured");
        }

        String body;
        try {
            body = buildBody(question);
        } catch (JSONException e) {
            throw new IOException("could not build the request: " + e.getMessage(), e);
        }

        return HttpRetry.execute(provider.label, new HttpRetry.Request<String>() {
            @Override
            public String execute() throws IOException {
                return askOnce(body);
            }
        });
    }

    private String askOnce(String body) throws IOException {
        URL url = HttpEndpoint.parse(endpoint(), provider.label + " endpoint");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        if (ignoreSsl) {
            com.myvu.client.core.SslUtils.applySslBypass(conn);
        }
        try {
            conn.setRequestMethod("POST");
            conn.setRequestProperty("content-type", "application/json");
            authorize(conn);
            conn.setConnectTimeout(TIMEOUT_MS);
            conn.setReadTimeout(TIMEOUT_MS);
            conn.setDoOutput(true);

            OutputStream out = conn.getOutputStream();
            out.write(body.getBytes(StandardCharsets.UTF_8));
            out.close();

            int status = conn.getResponseCode();
            String response = readAll(status >= 400 ? conn.getErrorStream() : conn.getInputStream());
            if (status >= 400) {
                throw HttpRetry.statusError(status, provider.label + " API returned " + status
                        + ": " + extractError(response));
            }

            String text;
            try {
                text = extractText(response).trim();
            } catch (JSONException e) {
                throw new IOException("unparseable " + provider.label + " response: "
                        + e.getMessage(), e);
            }
            if (text.isEmpty()) {
                throw new IOException(provider.label + " returned an empty answer");
            }
            return text;
        } finally {
            conn.disconnect();
        }
    }

    /** The supported JSON APIs put failures under error.message. */
    private static String extractError(String response) {
        try {
            JSONObject error = new JSONObject(response).optJSONObject("error");
            if (error != null) return error.optString("message", response);
        } catch (JSONException ignored) {
            // Fall through to the raw body.
        }
        return response.substring(0, Math.min(200, response.length()));
    }

    private static String readAll(InputStream in) throws IOException {
        if (in == null) return "";
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buf = new byte[8192];
        int n;
        while ((n = in.read(buf)) > 0) out.write(buf, 0, n);
        return new String(out.toByteArray(), StandardCharsets.UTF_8);
    }

    private static void disableSslVerification(javax.net.ssl.HttpsURLConnection conn) {
        try {
            javax.net.ssl.TrustManager[] trustAllCerts = new javax.net.ssl.TrustManager[]{
                new javax.net.ssl.X509TrustManager() {
                    public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                        return new java.security.cert.X509Certificate[0];
                    }
                    public void checkClientTrusted(java.security.cert.X509Certificate[] certs, String authType) {}
                    public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType) {}
                }
            };
            javax.net.ssl.SSLContext sc = javax.net.ssl.SSLContext.getInstance("TLS");
            sc.init(null, trustAllCerts, new java.security.SecureRandom());
            conn.setSSLSocketFactory(sc.getSocketFactory());
            conn.setHostnameVerifier((hostname, session) -> true);
        } catch (Exception e) {
            com.myvu.client.core.LogBus.error("could not disable SSL verification", e);
        }
    }
}

File: android-kotlin/app/src/main/java/com/myvu/client/ai/AiHttpClient.kt
package com.myvu.client.ai

import com.myvu.client.core.LogBus
import com.myvu.client.core.SslUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import javax.net.ssl.SSLException

/**
 * Shared plumbing for the provider clients: one JSON POST, one JSON answer.
 */
abstract class AiHttpClient @JvmOverloads constructor(
    protected val provider: AiProvider,
    protected val apiKey: String?,
    model: String?,
    systemPrompt: String?,
    protected val ignoreSsl: Boolean = false
) : AiClient {

    protected val model: String = if (model.isNullOrBlank()) provider.displayName else model.trim()
    protected val systemPrompt: String = if (systemPrompt.isNullOrBlank()) AiClient.DEFAULT_SYSTEM_PROMPT else systemPrompt.trim()

    override fun isConfigured(): Boolean {
        return !apiKey.isNullOrBlank()
    }

    protected abstract fun endpoint(): String
    protected abstract fun authorize(conn: HttpURLConnection)
    @Throws(JSONException::class)
    protected abstract fun buildBody(question: String): String
    @Throws(JSONException::class)
    protected abstract fun extractText(response: String): String

    @Throws(IOException::class)
    override fun ask(question: String): String {
        if (!isConfigured()) {
            throw IOException("${provider.displayName} is not fully configured")
        }

        val body = try {
            buildBody(question)
        } catch (e: JSONException) {
            throw IOException("could not build the request: ${e.message}", e)
        }

        return HttpRetry.execute(provider.displayName) {
            askOnce(body)
        }
    }

    @Throws(IOException::class)
    @JvmOverloads
    protected fun postRaw(body: String, customReadTimeoutMs: Int? = null): String {
        return postRawInternal(body, ignoreSsl, customReadTimeoutMs)
    }

    @Throws(IOException::class)
    private fun postRawInternal(body: String, bypassSsl: Boolean, customReadTimeoutMs: Int? = null): String {
        val targetUrl = endpoint()
        LogBus.log("${provider.displayName}: POST $targetUrl (payload: ${body.length} chars)")
        val url = HttpEndpoint.parse(targetUrl, "${provider.displayName} endpoint")
        val conn = url.openConnection() as HttpURLConnection
        if (bypassSsl) {
            SslUtils.applySslBypass(conn)
        }
        try {
            conn.requestMethod = "POST"
            conn.setRequestProperty("content-type", "application/json")
            authorize(conn)
            val isPrivateHost = targetUrl.contains("://10.") ||
                    targetUrl.contains("://192.168.") ||
                    targetUrl.contains("://172.16.") ||
                    targetUrl.contains("://172.17.") ||
                    targetUrl.contains("://172.18.") ||
                    targetUrl.contains("://172.19.") ||
                    targetUrl.contains("://172.2") ||
                    targetUrl.contains("://172.30.") ||
                    targetUrl.contains("://172.31.") ||
                    targetUrl.contains("localhost") ||
                    targetUrl.contains("127.0.0.1")
            val isLocal = isPrivateHost
            conn.connectTimeout = if (isLocal) LOCAL_CONNECT_TIMEOUT_MS else CONNECT_TIMEOUT_MS
            conn.readTimeout = customReadTimeoutMs ?: if (isLocal) LOCAL_READ_TIMEOUT_MS else READ_TIMEOUT_MS
            conn.doOutput = true

            conn.outputStream.use { out ->
                out.write(body.toByteArray(StandardCharsets.UTF_8))
            }

            val status = conn.responseCode
            val stream = if (status >= 400) conn.errorStream else conn.inputStream
            val response = readAll(stream)
            LogBus.log("${provider.displayName}: HTTP $status (${response.length} chars received)")
            if (status >= 400) {
                throw HttpRetry.statusError(
                    status,
                    "${provider.displayName} API returned $status: ${extractError(response)}"
                )
            }
            return response
        } catch (e: SSLException) {
            if (!bypassSsl) {
                LogBus.warn("${provider.displayName} SSL failed, retrying with SSL bypass...")
                return postRawInternal(body, true, customReadTimeoutMs)
            }
            throw e
        } finally {
            conn.disconnect()
        }
    }

    @Throws(IOException::class)
    protected fun askOnce(body: String): String {
        val response = postRaw(body)
        val text = try {
            extractText(response).trim()
        } catch (e: JSONException) {
            throw IOException("unparseable ${provider.displayName} response: ${e.message}", e)
        }
        if (text.isEmpty()) {
            throw IOException("${provider.displayName} returned an empty answer")
        }
        return text
    }

    companion object {
        private const val CONNECT_TIMEOUT_MS = 15000
        private const val READ_TIMEOUT_MS = 60000
        private const val LOCAL_CONNECT_TIMEOUT_MS = 15000
        private const val LOCAL_READ_TIMEOUT_MS = 30000

        private fun extractError(response: String): String {
            try {
                val error = JSONObject(response).optJSONObject("error")
                if (error != null) return error.optString("message", response)
            } catch (ignored: JSONException) {
            }
            return response.substring(0, Math.min(200, response.length))
        }

        private fun readAll(input: InputStream?): String {
            if (input == null) return ""
            return input.use { String(it.readBytes(), StandardCharsets.UTF_8) }
        }
    }
}
